Rou Zhang 7050809338

In [1]:
import numpy as np
import pandas as pd
from pandas import DataFrame,Series
import os
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
import statsmodels.api as sm
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.model_selection import cross_val_score
from sklearn import datasets
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import hamming_loss

/anaconda2/lib/python2.7/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [9]:
path='/Users/zhangrou/Desktop/courses/INF552/HW4/Frogs_MFCCs.csv'
f=pd.read_csv(path)

In [10]:
family_score={'Bufonidae':1,'Dendrobatidae':2,'Hylidae':3,'Leptodactylidae':4}
genus_score={'Adenomera':1,'Ameerega':2,'Dendropsophus':3,'Hypsiboas':4,'Leptodactylus':5,'Osteocephalus':6,'Rhinella':7,'Scinax':8}
species_score={
'AdenomeraAndre':1, 
'AdenomeraHylaedactylus':2,
'Ameeregatrivittata':3, 
'HylaMinuta':4, 
'HypsiboasCinerascens':5,
'HypsiboasCordobae':6,
'LeptodactylusFuscus':7,
'OsteocephalusOophagus':8, 
'Rhinellagranulosa':9, 
'ScinaxRuber':10
}
f['Family']=f['Family'].map(family_score)
f['Genus']=f['Genus'].map(genus_score)
f['Species']=f['Species'].map(species_score)
train_df, test_df = train_test_split(f, test_size=0.3)

b)i. Research exact match and hamming score/ loss methods for evaluating multi- label classification and use them in evaluating the classifiers in this problem.


Exact match: Exact match is most strict metric, indicating the percentage of samples that have all their labels classified correctly.


Hamming score: The Hamming loss is the fraction of labels that are incorrectly predicted.

In [13]:
from sklearn.svm import LinearSVC
from sklearn.datasets import make_classification
from sklearn import svm, datasets
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import LeaveOneOut
import numpy as np 
from sklearn.svm import SVC
parameters={'C':[0.01,0.1,1,10],'gamma':[0.1,0.5,1.0,1.5]}
svc = SVC(decision_function_shape='ovr')
clf = GridSearchCV(svc, parameters)
x=train_df.iloc[:,:22]
y_family=train_df['Family']
y_genus=train_df['Genus']
y_species=train_df['Species']

clf.fit(x,y_family)
params=clf.best_params_ 
print params
best_svc = SVC(C=params['C'],gamma=params['gamma'])
best_svc.fit(x,y_family)
family_score=best_svc.score(test_df.iloc[:,:22],test_df['Family'])
y_pred=best_svc.predict(test_df.iloc[:,:22])
hamming=hamming_loss(test_df['Family'], y_pred)
print hamming
#print family_score


clf.fit(x,y_genus)
params=clf.best_params_ 
print params
best_svc = SVC(C=params['C'],gamma=params['gamma'])
best_svc.fit(x,y_genus)
genus_score=best_svc.score(test_df.iloc[:,:22],test_df['Genus'])
y_pred=best_svc.predict(test_df.iloc[:,:22])
hamming=hamming_loss(test_df['Genus'], y_pred)
print hamming
#print genus_score

clf.fit(x,y_species)
params=clf.best_params_ 
print params
best_svc = SVC(C=params['C'],gamma=params['gamma'])
best_svc.fit(x,y_species)
species_score=best_svc.score(test_df.iloc[:,:22],test_df['Species'])
y_pred=best_svc.predict(test_df.iloc[:,:22])
hamming=hamming_loss(test_df['Species'], y_pred)
print hamming
#print species_score




{'C': 10, 'gamma': 1.5}
0.006947660954145438
{'C': 10, 'gamma': 1.5}
0.010189902732746642
{'C': 10, 'gamma': 1.5}
0.009726725335803613


b)ii. Train a SVM for each of the labels, using Gaussian kernels and one versus all classifiers. Determine the weight of the SVM penalty and the width of the Gaussian Kernel using 10 fold cross validation. You are welcome to try to solve the problem with both normalized and raw attributes and report the results.


In this step, I set parameters={'C':[0.01,0.1,1,10],'gamma':[0.1,0.5,1.0,1.5]} and used GridSearchCV to choose best parameters. 

For the Family lable, the best C is 10 and the best gamma is 1.5, the  hamming loss is  0.006947660954145438 

For the Genus lable, the best C is 10 and the best gamma is 1.5,the hamming loss is 0.010189902732746642

For the Species lable, the best C is 10 and the best gamma is 1.5,the hamming loss is 
0.009726725335803613

In [14]:
from sklearn import preprocessing 
norm_data=pd.DataFrame()
for i in range(22):
    min_max_scaler = preprocessing.MinMaxScaler() 
    data=pd.DataFrame(min_max_scaler.fit_transform(f.iloc[:,i].reshape(-1,1)))  
    norm_data=pd.concat([norm_data,data],axis=1)  
norm_data=pd.concat([norm_data,f.iloc[:,22:26]],axis=1)
norm_data.columns=f.columns
train_df, test_df = train_test_split(norm_data, test_size=0.3)

/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:5: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  """


In [15]:
x=train_df.iloc[:,:22]
y_family=train_df['Family']
y_genus=train_df['Genus']
y_species=train_df['Species']

parameters={'C':[0.01,0.1,1,10]}
svc = LinearSVC(penalty='l1',dual=False)
clf = GridSearchCV(svc, parameters)

clf.fit(x,y_family)
params=clf.best_params_ 
print params
best_svc = LinearSVC(C=params['C'],penalty='l1',dual=False)
best_svc.fit(x,y_family)
family_score=best_svc.score(test_df.iloc[:,:22],test_df['Family'])
print 1-family_score

clf.fit(x,y_genus)
params=clf.best_params_ 
print params
best_svc = LinearSVC(C=params['C'],penalty='l1',dual=False)
best_svc.fit(x,y_genus)
genus_score=best_svc.score(test_df.iloc[:,:22],test_df['Genus'])
print 1-genus_score

clf.fit(x,y_species)
params=clf.best_params_ 
print params
best_svc = LinearSVC(C=params['C'],penalty='l1',dual=False)
best_svc.fit(x,y_species)
species_score=best_svc.score(test_df.iloc[:,:22],test_df['Species'])
print 1-species_score


{'C': 10}
0.063918480778138
{'C': 10}
0.05743399722093567
{'C': 10}
0.044001852709587785


b)iii. Repeat 2(b)ii with L1-penalized SVMs. Remember to normalize the at- tributes.

In this step, I set parameters={'C':[0.01,0.1,1,10],'gamma':[0.1,0.5,1.0,1.5]} and used GridSearchCV to choose best parameters. 

For the Family lable, the best C is 10,the  hamming loss  is 0.063918480778138

For the Genus lable, the best C is 10 ,the  hamming loss  is 0.05743399722093567

For the Species lable, the best C is 10,the  hamming loss  is 0.044001852709587785

In [17]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import SMOTE 
from sklearn.svm import LinearSVC
from sklearn.datasets import make_classification
from sklearn import svm, datasets
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import LeaveOneOut
import numpy as np 
from sklearn.svm import SVC
sm = SMOTE(ratio="minority",random_state=42,kind='svm')
parameters={'C':[0.01,0.1,1,10]}
svc = LinearSVC(penalty='l1',dual=False)
clf = GridSearchCV(svc, parameters)

x=train_df.iloc[:,:22]
y_family=train_df['Family']
y_genus=train_df['Genus']
y_species=train_df['Species']

x, y_family = sm.fit_sample(x, y_family)
x=pd.DataFrame(x)
y_family=pd.DataFrame(y_family)
clf.fit(x,y_family)
params=clf.best_params_
print params
best_svc = LinearSVC(C=params['C'],penalty='l1',dual=False)
best_svc.fit(x,y_family)
family_score=best_svc.score(test_df.iloc[:,:22],test_df['Family'])
print 1-family_score
x=train_df.iloc[:,:22]
x, y_genus = sm.fit_sample(x, y_genus)
x=pd.DataFrame(x)
y_genus=pd.DataFrame(y_genus)
clf.fit(x,y_genus)
params=clf.best_params_ 
print params
best_svc = LinearSVC(C=params['C'],penalty='l1',dual=False)
best_svc.fit(x,y_genus)
genus_score=best_svc.score(test_df.iloc[:,:22],test_df['Genus'])
print 1-genus_score
x=train_df.iloc[:,:22]
x, y_species = sm.fit_sample(x, y_species)
x=pd.DataFrame(x)
y_species=pd.DataFrame(y_species)
clf.fit(x,y_species)
params=clf.best_params_ 
print params
best_svc = LinearSVC(C=params['C'],penalty='l1',dual=False)
best_svc.fit(x,y_species)
species_score=best_svc.score(test_df.iloc[:,:22],test_df['Species'])
print 1-species_score


{'C': 10}
0.08012968967114409
{'C': 10}
0.058360352014821704
{'C': 10}
0.050023158869847206


b)iv. Repeat 2(b)iii by using SMOTE or any other method you know to remedy class imbalance. Report your conclusions about the classifiers you trained.


In this step, I firstly used " SMOTE(ratio="minority")" function to compensate imbalanced classes. And this function will increase the number of minority to the same number of majority  class.

Then I set parameters={'C':[0.01,0.1,1,10],'gamma':[0.1,0.5,1.0,1.5]} and used GridSearchCV to choose best parameters. 

For the Family lable, the best C is 10,the  hamming loss  is 0.08012968967114409

For the Genus lable, the best C is 10 ,the  hamming loss  is 0.058360352014821704

For the Species lable, the best C is 10,the hamming loss  is 0.050023158869847206


We can see the result with using SMOTE in this step actually is very similar to the result we got in last step, so I think imbalanced classes doen't have a heavy influence on the model training. Maybe because the model only depends on  support vector machine .